proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,252 @@
+
+```
+#interface-instr ::=
+  | ..
+  | <numeric instructions>
+  | <memory instructions>
+  | <table instructions>
+  | drop
+  | unreachable
+  | nop
+```
+
+Refer to the definition of instruction classes [here](https://webassembly.github.io/spec/core/syntax/instructions.html).
+
+## Laziness
+
+Interface values generated by interface instructions are lazy. The execution of an instruction generating an interface value happens immediately before the instruction that consumes the value.
+
+## `bool.lift_iMM`
+
+This instruction has type `[iMM] -> [bool]`.
+
+The output is `true` iff the input `iMM` has any bit set in the integer.
+
+## `bool.lower_iMM`
+
+This instruction has type `[bool] -> iMM`.
+
+The output is `0b1` zero extended to the desired bitwidth iff the input is true.
+
+## `sNN.lift_iMM`
+
+This instruction has type `iMM` -> `sNN`.
+
+The output is the value of `iMM` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `sNN.lower_iMM`
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lift_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `uNN.lower_iMM`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction traps.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions traps if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding $alloc: #funcidx
+```
+
+The type of this instruction is:
+```
+[string] -> [i32, i32]
+
+Where:
+  * $alloc has a valid functype of [i32] -> [i32]
+```
+
+The instruction will execute the following steps:
+```
+  1. Let $length be the bytes required to encode the string with #encoding
+  2. Call $alloc with $length and receive an offset into #memidx
+  3. Encode the string into the offset at memory
+```
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
```

Comment:
use pack and unpack here.
pack takes a sequence of IT values and makes a single IT value.
This shows that there will be IT instructions that are not directly lowering or lifting.

Reply:
If I understand correctly, pack pops a sequence of IT values from the stack and produce a single IT record value. And unpack takes a single IT record value and pushes a sequence of IT values to the stack. This is exactly what intended for 'record.lift'/'record.lower' to do.

Is the issue with naming conventions as earlier discussed?